### PR TITLE
3.1 ratelimit source

### PR DESCRIPTION
**Pull request by OliviaYtterbrink**
Rate limit authentication attempts per source instead of per user.
This solved an issue where it was possible to lock out a user from the server, just by attempting to log in with wrong credentials.

changelog: Limit authentication requests depending on their originating source

### CODE REVIEW

File: community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/BoltConnectionIT.java [hunk 1]
```diff
@@ -133,8 +134,8 @@ public void shouldCloseConnectionOnPullAllBeforeInit() throws Throwable
     public void shouldExecuteStatement() throws Throwable
     {
         // Given
-        BoltStateMachine machine = env.newMachine( "<test>" );
-        machine.init( USER_AGENT, emptyMap(), null );
+        BoltStateMachine machine = env.newMachine( SOURCE, "<test>" );
+        machine.init( USER_AGENT, EMPTY_MAP, null );
```
Comment (fickludd) [inline community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/BoltConnectionIT.java:138]: Helper method for these two lines could save some boilerplate in this class.

File: community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java [hunk 2]
```diff
@@ -87,14 +87,27 @@ public AuthenticationResult authenticate( User user, String password)
...
+        if ( authMetaMap == null )
+        {
+            authMetaMap = new ConcurrentHashMap<>();
+            ConcurrentMap<String,AuthenticationMetadata> preExisting = authenticationData.putIfAbsent( username, authMetaMap );
+
+            if ( preExisting != null )
+            {
+                authMetaMap = preExisting;
+            }
+        }
```
Comment (fickludd) [inline community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java:103]: Instead of this dance, how about `ConcurrentMap.computeIfAbsent`? That would replace both `get` and `putIfAbsent` and guarantees concurrently nice behaviour.

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ThreadedTransaction.java [hunk 3]
```diff
@@ -102,22 +104,34 @@ public Throwable apply( S subject )
...
                         {
+                            if ( state == LatchState.NOT_STARTED )
+                            {
+                                latch.start();
+                                state = LatchState.STARTED;
+                            }
+                            if ( state == LatchState.STARTED )
+                            {
+                                latch.finish();
+                            }
```
Comment (fickludd) [inline enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ThreadedTransaction.java:134]: Eh, how is this related to rate limiting? Confused. And also I don't really understand when this code is supposed to kick in - the inner finally block ends with a `state = LatchState.FINISHED` so it should always be finished?

**Review by fickludd — COMMENTED** [review]
Hey folks, I randomly saw this PR and decided to have a look. 

I must say that I'm kinda sceptical about _changing_ this behaviour in a _patch release_. How about config + retained default behaviour in `3.1`+`3.2` => full monkey in `3.3`?

Also did you think about the fact that this change now allows attempts per source + user, allows me to scale my brute-force attempt with my drone computer farm size? This reduces the strength of the rate limit by a few orders of magnitude.

File: enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ThreadedTransaction.java (hunk 3 above)
Comment (OliviaYtterbrink) [reply enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ThreadedTransaction.java:134]: You're correct in part, there's some redundant checks in this block, since it will always be in one of two states. But it's required that the latch be counted down since otherwise the tests will hang if there's an error in the outer try block.

File: community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java (hunk 2 above)
Comment (OliviaYtterbrink) [reply community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java:103]: Good suggestion

File: community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/BoltConnectionIT.java (hunk 1 above)
Comment (OliviaYtterbrink) [reply community/bolt/src/test/java/org/neo4j/bolt/v1/runtime/integration/BoltConnectionIT.java:138]: +1

**Comment by OliviaYtterbrink** [comment]
Since this is such a big change, we've decided to close this PR and have a look at if this should go in at all and if so then it should be configurable.
